sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2163,22 +2235,29 @@ class Analyzer(
                   alias.toAttribute
               }
           }
-
-          // Push the aggregate expressions into the aggregate (if any).
-          if (aggregateExpressions.nonEmpty) {
-            Project(agg.output,
-              Filter(transformedAggregateFilter,
-                agg.copy(aggregateExpressions = agg.aggregateExpressions ++ aggregateExpressions)))
-          } else {
-            filter
-          }
+          (aggregateExpressions, Some(transformedAggregateFilter))
         } else {
-          filter
+          (aggregateExpressions, None)
         }
       } catch {
         // Attempting to resolve in the aggregate can result in ambiguity.  When this happens,
         // just return the original plan.
-        case ae: AnalysisException => filter
+        case ae: AnalysisException => (aggregateExpressions, None)
+      }
+    }
+
+    def resolveHaving(filter: Filter, agg: Aggregate): LogicalPlan = {
+      // Try resolving the condition of the filter as though it is in the aggregate clause
+      val (aggregateExpressions, resolvedHavingCond) =
+        resolveFilterCondInAggregate(filter.condition, agg)
+
+      // Push the aggregate expressions into the aggregate (if any).
+      if (aggregateExpressions.nonEmpty) {
```

Comment:
you seem to assume that, if `aggregateExpressions.nonEmpty`, then `resolvedHavingCond` must be `Some`. Let's make the `resolveFilterCondInAggregate` more clear with this assumption.

Reply:
Copy, done in 3b48e38.